pulp_rpm/app/downloaders.py

```diff
@@ -94,3 +104,88 @@ async def _run(self, extra_data=None):
         if self._close_session_on_finalize:
             self.session.close()
         return to_return
+
+
+class UlnDownloader(RpmDownloader):
+    """
+    Custom Downloader for ULN repositories.
+
+    Args:
+        uln_user (str): Username for authentification in ULN network
+        uln_pwd (str): password for authentification in ULN network
+        server_url (str): ULN server url. If no url is given, use default oracle
+            ULN server url `https://linux-update.oracle.com/`
+
+    Raises:
+        NoUlnCredentials: If no or not valid ULN credentials are given, this Error will be displayed
+    """
+
+    def __init__(self, *args, uln_user=None, uln_pwd=None, server_url=None, **kwargs):
+        """
+        Initialize the downloader for ULN repositories.
+
+        If no server URL is given, use the server url from oracle linux.
+        """
+        self.uln_user = uln_user
+        self.uln_pwd = uln_pwd
+        if server_url:
+            self.server_url = server_url
+        else:  # if no server_url is given, use oracle linux url
+            self.server_url = "https://linux-update.oracle.com/"
+        self.headers = None
+        self.session_key = None
+
+        super().__init__(*args, **kwargs)
+
+    async def _run(self, extra_data=None):
+        """
+        Download, validate, and compute digests on the `url`. This is a coroutine.
+
+        Once per session the coroutine logs into the ULN account using the
+        uln_username and uln_pwd. Ther returned key is used for authentification
+        for all other downloads.
+
+        This method provides the same return object type and documented in
+        :meth:`~pulpcore.plugin.download.BaseDownloader._run`.
+        """
+        if self.sles_auth_token:
+            auth_param = f"?{self.sles_auth_token}"
+            url = urljoin(self.url, auth_param)
```

Comment:
I don't think we need to account for SLES auth tokens in a downloader specifically for ULN type repos. I don't think there will ever be a situation were some remote repository requires multiple authentication mechanisms (and if there is, that will be a new story for future developers)...

Reply:
remove it. good point @quba42

Reply:
Done